File: core/src/rustyice/game/physics/SBPhysicsComponent.java
package rustyice.game.physics;

import com.badlogic.gdx.math.MathUtils;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.physics.box2d.*;
import com.badlogic.gdx.physics.box2d.BodyDef.BodyType;

import rustyice.editor.annotations.ComponentProperty;
import rustyice.game.GameObject;
import rustyice.physics.Collision;
import rustyice.physics.PhysicsComponent;

/**
 * @author gabek Single body physics component
 */
public class SBPhysicsComponent implements PhysicsComponent {

    protected transient Body body;
    private transient boolean initialized = false;

    private BodyType bodytype = BodyType.DynamicBody;
    private boolean flying;
    private float groundFriction = .5f;
    private float x, y, rotation;
    private GameObject parent;

    public SBPhysicsComponent() {
    }

    public GameObject getParent(){
        return parent;
    }

    @Override
    public void init(GameObject parent) {
        if (!initialized) {
            initialized = true;
            this.parent = parent;

            BodyDef bodyDef = buildBodyDef();

            body = parent.getSection().getWorld().createBody(bodyDef);
        }
    }
    
    private void updateDrag() {
        if (initialized && !flying) {
            body.setLinearDamping(body.getMass() * groundFriction);
            body.setAngularDamping(body.getMass() * groundFriction);
        }
    }

    public boolean isInitialized() {
        return initialized;
    }

    @ComponentProperty(title = "Flying")
    public void setFlying(boolean flying) {
        this.flying = flying;
        if (isInitialized()) {
            if (flying) {
                body.setLinearDamping(0);
            } else {
                body.setLinearDamping(body.getMass() * groundFriction);
            }
        }
    }

    @ComponentProperty(title = "Flying")
    public boolean isFlying(){
        return flying;
    }

    @Override
    public void update(float delta) {
        rotation = MathUtils.radiansToDegrees * body.getAngle();
        x = body.getPosition().x;// - master.getWidth()/2;
        y = body.getPosition().y;// - master.getHeight()/2;

        if (!flying && Math.abs(body.getLinearVelocity().x) < .1f && Math.abs(body.getLinearVelocity().y) < .1f && !body.getLinearVelocity().isZero()) {
            body.setLinearVelocity(Vector2.Zero);
        }
    }

    @Override
    public void beginCollision(Collision collision) {

    }

    @Override
    public void endCollision(Collision collision) {

    }

    @Override
    public void store() {
        if (initialized) {
            initialized = false;
            parent.getSection().getWorld().destroyBody(body);
            body = null;
        }
    }

    public Body getBody() {
        return body;
    }

    protected BodyDef buildBodyDef() {
        BodyDef bodyDef = new BodyDef();

        
        bodyDef.type = bodytype;
        // bodyDef.fixedRotation = true;

        bodyDef.position.x = x;
        bodyDef.position.y = y;
        bodyDef.angle = rotation * MathUtils.degRad;

        return bodyDef;
    }
    
    public Fixture addRectangle(float width, float height, FixtureDef fixtureDef) {
        PolygonShape rect = new PolygonShape();
        rect.setAsBox(width / 2, height / 2);

        fixtureDef.shape = rect;
        Fixture fix = body.createFixture(fixtureDef);
        rect.dispose();

        fix.setUserData(parent);
        updateDrag();
        return fix;
    }

    public Fixture addCircle(float radius, FixtureDef fixtureDef) {
        CircleShape circle = new CircleShape();
        circle.setRadius(radius);

        fixtureDef.shape = circle;
        Fixture fix = body.createFixture(fixtureDef);
        circle.dispose();

        fix.setUserData(parent);
        updateDrag();
        return fix;
    }

    @Override
    public float getX() {
        return x;
    }

    @Override
    public float getY() {
        return y;
    }

    @Override
    public float getRotation() {
        return rotation;
    }

    @Override
    public void setX(float x) {
        this.x = x;
        if (initialized) {
            body.setTransform(x, body.getPosition().y, body.getAngle());
        }
    }

    @Override
    public void setY(float y) {
        this.y = y;
        if (initialized) {
            body.setTransform(body.getPosition().x, y, body.getAngle());
        }
    }

    @Override
    public void setRotation(float rotation) {
        this.rotation = rotation;
        if (initialized) {
            body.setTransform(body.getPosition(), rotation * MathUtils.degRad);
        }
    }

    @Override
    public void setPosition(float x, float y) {
        this.x = x;
        this.y = y;
        if (initialized) {
            body.setTransform(x, y, body.getAngle());
        }
    }
}

File: core/src/rustyice/game/physics/SBPhysicsComponent.kt
package rustyice.game.physics

import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.math.MathUtils
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.physics.box2d.*
import com.badlogic.gdx.physics.box2d.BodyDef.BodyType
import rustyice.editor.annotations.ComponentProperty
import rustyice.graphics.Camera
import rustyice.physics.Collision
import rustyice.physics.PhysicsComponent

/**
 * Single body physics component
 * @author gabek
 */
open class SBPhysicsComponent: PhysicsComponent() {
    @Transient protected var body: Body? = null
        private set

    var bodyType = BodyType.DynamicBody

    @ComponentProperty
    var flying = false
        set(value) {
            field = value
            body?.let{ it.linearDamping = if(flying) 0f  else it.mass * groundFriction}
        }

    var groundFriction = .5f

    private var _x = 0f
    private var _y = 0f
    private var _rotation = 0f

    override var x: Float
        get() = body?.position?.x ?: _x
        set(value) {
            val body = body
            _x = value
            body?.setTransform(value, body.position.y, body.transform.rotation)
        }

    override var y: Float
        get() = body?.position?.y ?: _y
        set(value) {
            val body = body
            _y = value
            body?.setTransform(body.position.x, value, body.angle)
        }

    override var rotation: Float
        get() = (body?.transform?.rotation ?: _rotation) * MathUtils.radDeg
        set(value) {
            val body = body
            _rotation = value
            body?.setTransform(body.position, value * MathUtils.degRad)
        }

    override fun init() {
        super.init()
        val world = section!!.game!!.world

        val bodyDef = buildBodyDef()

        body = world.createBody(bodyDef)
    }
    
    private fun updateDrag() {
        val body = body

        if (isInitialized && !flying && body != null) {
            body.linearDamping = body.mass * groundFriction
            body.angularDamping = body.mass * groundFriction
        }
    }

    override fun update(delta: Float) {
        val body = body

        if(body != null) {
            if (!flying && Math.abs(body.linearVelocity.x) < .1f && Math.abs(body.linearVelocity.y) < .1f && !body.linearVelocity.isZero) {
                _x = body.position.x
                _y = body.position.y
                _rotation = body.transform.rotation

                body.linearVelocity = Vector2.Zero
            }
        }
    }

    override fun beginCollision(collision: Collision) {}

    override fun endCollision(collision: Collision) {}

    override fun store() {
        super.store()
        val body = body

        if(body != null){
            parent!!.section!!.game!!.world.destroyBody(body)
            this.body = null
        }
    }

    protected fun buildBodyDef(): BodyDef {
        val bodyDef = BodyDef()

        
        bodyDef.type = bodyType
        // bodyDef.fixedRotation = true;

        bodyDef.position.x = _x
        bodyDef.position.y = _y
        bodyDef.angle = _rotation

        return bodyDef
    }
    
    fun addRectangle(width: Float, height: Float, fixtureDef: FixtureDef): Fixture {
        val body = body!!

        val rect = PolygonShape()
        rect.setAsBox(width / 2, height / 2)

        fixtureDef.shape = rect
        val fix = body.createFixture(fixtureDef)
        rect.dispose()

        fix.userData = parent
        updateDrag()
        return fix
    }

    fun addCircle(radius: Float, fixtureDef: FixtureDef): Fixture {
        val body = body!!

        val circle = CircleShape()
        circle.radius = radius

        fixtureDef.shape = circle
        val fix = body.createFixture(fixtureDef)
        circle.dispose()

        fix.userData = parent
        updateDrag()
        return fix
    }

    override fun render(batch: Batch, camera: Camera) {}
}
